### ISSUE TITLE
Broken move semantics in `decl_macro`

### ISSUE THREAD

**Issue by valff** [rust-lang/rust#46314]
Move semantics with declarative macros 2.0 are broken. Can copy non-copy types.
[Playground](https://play.rust-lang.org/?gist=24de21ef023ec4d6a08a76fbd507b68c&version=nightly)

```rust
#![feature(decl_macro)]
#![allow(unused_macros)]

#[derive(Debug)]
struct NonCopy(String);

struct Foo {
  x: NonCopy,
}

macro copy($foo:ident) {
  $foo.x
}

macro_rules! copy2 {
  ($foo:ident) => {
    $foo.x
  }
}

fn main() {
  let foo = Foo { x: NonCopy("foo".into()) };
  assert_two_copies(copy!(foo), foo.x);
  // A similar old-style macro will not work:
  //assert_two_copies(copy2!(foo), foo.x);
}

fn assert_two_copies(a: NonCopy, b: NonCopy) {
  println!("Got two copies: {:?}, {:?}", a, b);
}
```

rustc --version --verbose:

```
rustc 1.23.0-nightly (63739ab7b 2017-11-21)
binary: rustc
commit-hash: 63739ab7b210c1a8c890c2ea5238a3284877daa3
commit-date: 2017-11-21
host: x86_64-unknown-linux-gnu
release: 1.23.0-nightly
LLVM version: 4.0
```

**Comment by kennytm** [comment]
The error is caught when enabling MIR borrow-check (pass `-Z borrowck-mir` or `-Z nll`), so this only is a HIR borrow-check issue.

**Comment by petrochenkov** [comment]
Probably some fields are compared by name without hygiene adjustments.

**Comment by pnkfelix** [comment]
(tagging with nll labels just so we remember to look at this case when double checking lexical vs NLL.)

**Comment by nikomatsakis** [comment]
We should file this under https://github.com/rust-lang/rust/issues/47366, seems like?

**Comment by petrochenkov** [comment]
It's very likely this is fixed on my field hygiene branch, I'll check.

**Comment by valff** [comment]
@nikomatsakis Thanks for the reference. I will make a PR with a test for this bug soon.
